tokens-accounting/src/data.rs

```diff
@@ -0,0 +1,167 @@
+}
+
+#[derive(Clone, Encode, Decode, Debug, PartialEq, Eq)]
+pub struct TokensAccountingDeltaData {
+    pub(crate) token_data: DeltaDataCollection<TokenId, TokenData>,
+    pub(crate) circulating_supply: DeltaAmountCollection<TokenId>,
+}
+
+impl TokensAccountingDeltaData {
+    pub fn new() -> Self {
+        Self {
+            token_data: DeltaDataCollection::new(),
+            circulating_supply: DeltaAmountCollection::new(),
+        }
+    }
+
+    pub fn merge_with_delta(
+        &mut self,
+        other: TokensAccountingDeltaData,
+    ) -> Result<TokensAccountingDeltaUndoData, Error> {
+        let token_data_undo = self.token_data.merge_delta_data(other.token_data)?;
+
+        let circulating_supply_undo = other.circulating_supply.clone();
+        self.circulating_supply.merge_delta_amounts(other.circulating_supply)?;
+
+        Ok(TokensAccountingDeltaUndoData {
+            token_data: token_data_undo,
+            circulating_supply: circulating_supply_undo,
+        })
+    }
+}
+
+#[derive(Clone, Encode, Decode, Debug, PartialEq, Eq)]
+pub struct TokensAccountingDeltaUndoData {
+    pub(crate) token_data: DeltaDataUndoCollection<TokenId, TokenData>,
+    pub(crate) circulating_supply: DeltaAmountCollection<TokenId>,
+}
+
+impl TokensAccountingDeltaUndoData {
+    pub fn new() -> Self {
+        Self {
+            token_data: DeltaDataUndoCollection::new(),
+            circulating_supply: DeltaAmountCollection::new(),
+        }
+    }
+}
+
+#[derive(Clone, Encode, Decode, Debug, PartialEq, Eq)]
+pub enum TokenData {
+    FungibleToken(FungibleTokenData),
+}
+
+#[derive(Clone, Encode, Decode, Debug, PartialEq, Eq)]
+pub struct FungibleTokenData {
+    token_ticker: Vec<u8>,
+    number_of_decimals: u8,
+    metadata_uri: Vec<u8>,
+    total_supply: TokenTotalSupply,
+    locked: bool,
+    reissuance_controller: Destination,
+}
+
+impl FungibleTokenData {
+    pub fn new(
+        token_ticker: Vec<u8>,
+        number_of_decimals: u8,
+        metadata_uri: Vec<u8>,
+        total_supply: TokenTotalSupply,
+        locked: bool,
+        reissuance_controller: Destination,
+    ) -> Self {
+        Self {
+            token_ticker,
+            number_of_decimals,
+            metadata_uri,
+            total_supply,
+            locked,
+            reissuance_controller,
+        }
+    }
+
+    pub fn token_ticker(&self) -> &[u8] {
+        self.token_ticker.as_ref()
+    }
+
+    pub fn number_of_decimals(&self) -> u8 {
+        self.number_of_decimals
+    }
+
+    pub fn metadata_uri(&self) -> &[u8] {
+        self.metadata_uri.as_ref()
+    }
+
+    pub fn total_supply(&self) -> &TokenTotalSupply {
+        &self.total_supply
+    }
+
+    pub fn is_locked(&self) -> bool {
+        self.locked
+    }
+
+    pub fn reissuance_controller(&self) -> &Destination {
+        &self.reissuance_controller
+    }
+
+    pub fn lock(self) -> Option<Self> {
+        match self.total_supply {
```

Comment:
If locking fails, `self` is consumed and lost. Should it be returned in the `Err` arm of a result instead? Caller can `.ok()` it as needed. Also may be more fitting to call this `try_lock`.